core/core.schema.go

```diff
@@ -96,11 +111,58 @@ func (r *coreSchema) host(p graphql.ResolveParams) (any, error) {
 	return struct{}{}, nil
 }
 
+type Image struct {
```

Comment:
👍 Worth a try. I think that's the architecture we'll want in the future. If we can do it now, it moves us one step closer to #163

Sorry about the scope creep :) 

#163 has more details, but basically we're "in between" APIs right now

Reply:
Makes sense!

Is `ContainerID` going to be content-addressed like the other IDs? The comment makes it sound more like a UUID which would make it a pretty major departure from `FSID`. Maybe it's unique in order to support creating mutliple instances of the same container?

https://github.com/dagger/cloak/blob/63498e583021b15c2352574a8a7952221e1346a8/api/container.gql#L9-L10

Some trade-offs I see with embedding the image config in `FSID`:

1. Pro: we could add Dockerfile `ENV` and `DIR` equivalent operations to `Filesystem` for setting image config for a future exec
2. Pro: mirroring could "just work" since we'd be carrying around the OCI config verbatim regardless of whether Dagger knows or respects them
4. Con: becomes sort of a schema-within-a-schema. This is already technically true for `FSID` but that felt more like a case of embedding an externally generated black-box value, whereas here we'd be introducing our own layer around two external schemas, and one of them would be mutated directly by Dagger for the first point, so not really a black-box anymore. I don't think this is really a big deal practically, but it feels a little weird for an API where I'm otherwise biased towards keeping every exposed knob represented in a more regular fashion.
5. Con: we'd start using an "image-like" schema for things like output mounts which aren't really "image-like"
6. Con: longer `FSID`s?

I'm pretty mixed here but slightly biased towards `Image`, without really knowing the full implications of that yet.

The first "pro" could apply to `Image` too:

```graphql
type Image {
  fs: Filesystem!

  execInput: ExecInput!

  "Returns an Image with the env var set in execInput."
  withEnv(name: String!, value: String!): Image!

  "Pushes the image with the execInput as the image config."
  push(ref: String!): Boolean!
}
```

(These changes probably aren't worth pursuing until we figure out how it squares with the new API.)

The main difference here is avoiding the embedded wrapper schema and having a distinction between image-like filesystems and plain filesystems like output mounts. With these changes I could take an output rootfs mount (assuming I ran some command to build or unpack an image) and then provide image config for pushing it to a registry.

re: the second "pro", I'm not sure we'd be able to support digest-for-digest mirroring anyway, so maybe we don't care about that. There are plenty of other tools for that, so Dagger seems like an odd choice there.

Reply:
I'll take a deeper look at #163 soon and start making this change if not just for the sake of comparison. :slightly_smiling_face: